ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -3199,18 +3199,13 @@ private RelNode genFilterRelNode(ASTNode filterNode, RelNode srcRel,
     }
 
     private RelNode genFilterRelNode(RexNode filterExpression, RelNode srcRel,
-        ImmutableMap<String, Integer> outerNameToPosMap, RowResolver outerRR) throws SemanticException {
+        ImmutableMap<String, Integer> outerNameToPosMap, RowResolver outerRR) {
       if (RexUtil.isLiteral(filterExpression, false)
           && filterExpression.getType().getSqlTypeName() != SqlTypeName.BOOLEAN) {
```

Comment:
`WHERE 'foo' and 'bar'` is already supported through CBO because it goes through `TypeCheckProcFactory.DefaultExprProcessor#process()` as it tries to generate for the `and` tree. Because `and`'s children should be of type `boolean`, they get cast appropriately.

Whereas, when we only have `WHERE 'foo'`, it goes through `TypeCheckProcFactory.StrExprProcessor#process()`, which just returns the string back, without adding a cast.

Reply:
Thanks for the clarification. 

Can you please check if we support other non-boolean expressions in the WHERE clause. For instance:
```sql
WHERE t.stringColumn
WHERE t.intColumn
WHERE t.dateColumn
WHERE 'foo' || 'bar'
```
I would like to understand if both conditions of the `if` clause are necessary. In other words if we are about to introduce a `CAST` to `BOOLEAN` do we really care if the `filterExpression` is a literal?

Reply:
`WHERE <column>` goes through `TypeCheckProcFactory.ColumnExprProcessor#process()` and it looks like everything works well. There is already a file to test this: [vector_empty_where.q](https://github.com/apache/hive/blob/1c9969a003b09abc851ae7e19631ad208d3b6066/ql/src/test/queries/clientpositive/vector_empty_where.q).

For the concat operator,`WHERE 'foo' || 'bar'`, we get an empty result:
```
CBO PLAN:
HiveValues(tuples=[[]])
```

And for `where 'foo' or 'bar'`, the result is same as that with `and`.

I think we do need the literal check here because that is the only check that we don't do earlier than this point.

Reply:
The following expressions have the same type:
```sql
WHERE t.stringColumn
WHERE 'foo'
```
and result in a `WHERE` clause (`HiveFilter` operator) with a condition that is not `BOOLEAN`.

It's a bit weird that in the case that we have a `STRING` literal we need to add a `CAST` but when it is a `STRING` column we don't. 

What's preventing us from handling all non-boolean expressions uniformly? Are there problems if we always/never CAST?

Reply:
In Calcite's `ReduceExpressionsRule.FilterReduceExpressionsRule#onMatch()`, there's this code snippet:
```
      if (newConditionExp.isAlwaysTrue()) {
        call.transformTo(
            filter.getInput());
      } else if (newConditionExp instanceof RexLiteral
          || RexUtil.isNullLiteral(newConditionExp, true)) {
        call.transformTo(createEmptyRelOrEquivalent(call, filter));
      } else if (reduced) {
        call.transformTo(call.builder()
            .push(filter.getInput())
            .filter(newConditionExp).build());
      } else {
        if (newConditionExp instanceof RexCall) {
          boolean reverse = newConditionExp.getKind() == SqlKind.NOT;
          if (reverse) {
            newConditionExp = ((RexCall) newConditionExp).getOperands().get(0);
          }
          reduceNotNullableFilter(call, filter, newConditionExp, reverse);
        }
        return;
      }

```

For `WHERE t.stringColumn` the `newConditionExp` is a `RexInputRef` and we simply `return;` from the else block. Whereas, for `WHERE 'foo'` we go into the first `else if` block, which transforms the call to an empty `HiveValues`.

We can see the plans in the logs:

```
2024-06-25T14:29:53,756 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] parse.CalcitePlanner: Plan after decorrelation:
HiveProject(ctinyint=[$0], csmallint=[$1], cint=[$2], cbigint=[$3], cfloat=[$4], cdouble=[$5], cstring1=[$6], cstring2=[$7], ctimestamp1=[$8], ctimestamp2=[$9], cboolean1=[$10], cboolean2=[$11])
  HiveFilter(condition=[_UTF-16LE'foo':VARCHAR(2147483647) CHARACTER SET "UTF-16LE"])
    HiveTableScan(table=[[default, alltypesorc]], table:alias=[alltypesorc])

2024-06-25T14:29:53,756 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] log.PerfLogger: </PERFLOG method=Decorrelation start=1719350993756 end=1719350993756 duration=0 from=org.apache.hadoop.hive.ql.parse.CalcitePlanner$CalcitePlannerAction>
2024-06-25T14:29:53,756 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] log.PerfLogger: <PERFLOG method=Validate Query Materialization from=org.apache.hadoop.hive.ql.parse.CalcitePlanner$CalcitePlannerAction>
2024-06-25T14:29:53,756 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] log.PerfLogger: </PERFLOG method=Validate Query Materialization start=1719350993756 end=1719350993756 duration=0 from=org.apache.hadoop.hive.ql.parse.CalcitePlanner$CalcitePlannerAction>
2024-06-25T14:29:53,756 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] log.PerfLogger: <PERFLOG method=Calcite: Prejoin ordering transformation from=org.apache.hadoop.hive.ql.parse.CalcitePlanner$CalcitePlannerAction>
2024-06-25T14:29:53,770 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] AbstractRelOptPlanner.rule_execution_summary: Rule Attempts Info for HepPlanner
2024-06-25T14:29:53,770 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] AbstractRelOptPlanner.rule_execution_summary:
Rules                                                                   Attempts           Time (us)
HiveHepExtractRelNodeRule                                                      1                   2
* Total                                                                        1                   2

2024-06-25T14:29:53,780 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] rules.RelFieldTrimmer: Plan after trimming unused fields
HiveProject(ctinyint=[$0], csmallint=[$1], cint=[$2], cbigint=[$3], cfloat=[$4], cdouble=[$5], cstring1=[$6], cstring2=[$7], ctimestamp1=[$8], ctimestamp2=[$9], cboolean1=[$10], cboolean2=[$11])
  HiveValues(tuples=[[]])

2024-06-25T14:29:53,780 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] AbstractRelOptPlanner.rule_execution_summary: Rule Attempts Info for HepPlanner
2024-06-25T14:29:53,780 DEBUG [3a9dcba4-dfa5-4a29-b916-be29b1f6eab0 main] AbstractRelOptPlanner.rule_execution_summary:
Rules                                                                   Attempts           Time (us)
HiveFieldTrimmerRule                                                           2               9,647
ReduceExpressionsRule(Project)                                                 1              12,683
ReduceExpressionsRule(Filter)                                                  1                 419
HivePruneEmptyProject                                                          1                  85
HivePointLookupOptimizerRule(ProjectionExpressions)                            1                   9
```

Reply:
The fact that we reduce any kind of literal to false/empty seems like a bug in the rule and this behavior is not inline with the result of `RexLiteral#isAlwaysFalse` method. I think its safer to leave the literal as is at this stage. Please log a CALCITE ticket for this issue.

So based on the current situation it seems that we cannot avoid the CAST of literals till the respective CALCITE fix lands. (We don't have to wait for this to get fixed but its good to keep track of it). 

On the other hand, from a runtime perspective it seems impossible to evaluate the `WHERE t.stringColumn` clause without first casting to boolean (and the same goes for other non-boolean datatypes). I assume that this cast is added somehow by the (physical?) compiler. If that's the case then it seems reasonable to have this CAST also in the CBO plan for keeping things uniform and semantically sound. Although not formally forbidden, having non-boolean `RexNode` as conditions in a `Filter` operator is weird and seems kinda like a bug.

Reply:
Logged https://issues.apache.org/jira/browse/CALCITE-6448

Reply:
Removed the `isLiteral` check, which caused some changes to `vector_empty_where.q.out` - as seen in the latest commit.

I think the predicate changes make sense - kind of, but let me know what you think about them.

Reply:
The changes do seem like an improvement so we are good!